ocean_provider/validation/algo.py

```diff
@@ -157,3 +197,134 @@ def validate_formatted_algorithm_dict(algorithm_dict, algorithm_did):
         )  # noqa
 
     return True, ""
+
+
+class InputItemValidator(WorkflowValidator):
+    def __init__(self, consumer_address, provider_wallet, data, index):
+        """Initializes the input item validator."""
+        self.consumer_address = consumer_address
+        self.provider_wallet = provider_wallet
+        self.data = data
+        self.index = index
+
+    def validate(self):
+        required_keys = ["documentId", "transferTxId", "serviceId"]
+
+        for req_item in required_keys:
+            if not self.data.get(req_item) and not (
+                req_item == "serviceId" and self.data.get(req_item) == 0
+            ):
+                self.error = f"No {req_item} in input item."
+                return False
+
+        self.did = self.data.get("documentId")
+        try:
+            self.asset = get_asset_from_metadatastore(get_metadata_url(), self.did)
+        except ValueError:
+            self.error = f"Asset for did {self.did} not found."
+            return False
+
+        self.service = get_service_at_index(self.asset, self.data["serviceId"])
+
+        if not self.service:
+            self.error = f"Service index {self.data['serviceId']} not found."
+            return False
+
+        if self.service.type not in [
+            ServiceTypes.ASSET_ACCESS,
+            ServiceTypes.CLOUD_COMPUTE,
+        ]:
+            self.error = "Services in input can only be access or compute."
+            return False
+
+        asset_urls = get_asset_download_urls(
+            self.asset, self.provider_wallet, config_file=app.config["CONFIG_FILE"]
+        )
+
+        if self.service.type == ServiceTypes.CLOUD_COMPUTE and not asset_urls:
+            self.error = "Services in input with compute type must be in the same provider you are calling."
+            return False
+
+        if self.service.type != ServiceTypes.CLOUD_COMPUTE and self.index == 0:
+            self.error = "Service for main asset must be compute."
+            return False
```

Comment:
since this check does not rely on asset_urls, it should occur before getting asset_urls.